File: src/services/quota_service.py
from src.models.billing import Subscription, Plan
from src.core.redis import redis_client
from sqlalchemy.orm import Session
from datetime import datetime
from loguru import logger

class QuotaService:
    """
    Gerenciador de limites e quotas (SaaS Governing).
    Controla mensagens enviadas e instâncias de bots por Tenant.
    Replica o 'ResourceGuard' do .NET.
    """
    
    @staticmethod
    async def increment_message_usage(db: Session, tenant_id: str) -> bool:
        """
        Incrementa o contador de mensagens enviadas no mês.
        Retorna False se o limite do plano for atingido.
        """
        # 1. Recupera Plano do Tenant
        sub = db.query(Subscription).filter(Subscription.tenant_id == tenant_id).first()
        if not sub or sub.status != "active":
            return False
            
        plan = sub.plan
        month_key = f"usage:{tenant_id}:messages:{datetime.utcnow().strftime('%Y-%m')}"
        
        # 2. Verifica Limite no Redis (Caminho Rápido)
        current_count = await redis_client.get(month_key)
        current_count = int(current_count) if current_count else 0
        
        if current_count >= plan.max_messages_month:
            logger.warning(f"🚨 Tenant {tenant_id} atingiu cota de mensagens ({plan.max_messages_month})")
            return False
            
        # 3. Incrementa (Redis Atômico)
        await redis_client.incr(month_key)
        
        # TODO: Persistência assíncrona para Postgres em batch
        return True

    @staticmethod
    def can_create_bot(db: Session, tenant_id: str) -> bool:
        """Verifica se o Tenant ainda pode criar novas instâncias de bot."""
        from src.models.whatsapp import WhatsAppInstance
        
        sub = db.query(Subscription).filter(Subscription.tenant_id == tenant_id).first()
        if not sub: return False
        
        current_bots = db.query(WhatsAppInstance).filter(
            WhatsAppInstance.tenant_id == tenant_id,
            WhatsAppInstance.is_active == True
        ).count()
        
        return current_bots < sub.plan.max_bots

    @staticmethod
    def can_create_agent(db: Session, tenant_id: str) -> bool:
        """Verifica se o Tenant ainda pode criar novos usuários/agentes."""
        from src.models.user import User
        
        sub = db.query(Subscription).filter(Subscription.tenant_id == tenant_id).first()
        if not sub: return False
        
        current_agents = db.query(User).filter(
            User.tenant_id == tenant_id,
            User.is_agent == True
        ).count()
        
        return current_agents < sub.plan.max_agents

File: node-version/src/services/billing/quotaService.js
const { Subscription, WhatsAppInstance, User } = require('../../models/sql/models');
const redisClient = require('../../config/redis');
const logger = require('../../utils/logger');
const moment = require('moment');

class QuotaService {

  static async incrementMessageUsage(tenantId) {
    const sub = await Subscription.findOne({
      where: { tenant_id: tenantId },
      include: ['Plan'],
      ignoreTenant: true
    });

    if (!sub || sub.status !== 'active') return false;

    const monthKey = `usage:${tenantId}:messages:${moment().format('YYYY-MM')}`;
    
    let currentCount = await redisClient.get(monthKey);
    currentCount = currentCount ? parseInt(currentCount) : 0;

    if (currentCount >= sub.Plan.max_messages_month) {
      logger.warn(`🚨 Tenant ${tenantId} atingiu cota de mensagens (${sub.Plan.max_messages_month})`);
      return false;
    }

    await redisClient.incr(monthKey);
    return true;
  }

  static async canCreateBot(tenantId) {
    const sub = await Subscription.findOne({
      where: { tenant_id: tenantId },
      include: ['Plan'],
      ignoreTenant: true
    });
    if (!sub) return false;

    const currentBots = await WhatsAppInstance.count({
      where: { tenant_id: tenantId, is_active: true },
      ignoreTenant: true
    });

    return currentBots < sub.Plan.max_bots;
  }

  static async canCreateAgent(tenantId) {
    const sub = await Subscription.findOne({
      where: { tenant_id: tenantId },
      include: ['Plan'],
      ignoreTenant: true
    });
    if (!sub) return false;

    const currentAgents = await User.count({
      where: { tenant_id: tenantId, is_agent: true },
      ignoreTenant: true
    });

    return currentAgents < sub.Plan.max_agents;
  }
}

module.exports = QuotaService;
